### ISSUE TITLE
More idiomatic JSON format

### ISSUE THREAD

**Issue by DanielSWolf** [MontrealCorpusTools/Montreal-Forced-Aligner#453]
**Is your feature request related to a problem? Please describe.**

I'm very happy that recent MFA release candidates support JSON output. However, as someone who works with JSON data a lot, I find some aspects of the schema not quite idiomatic. So I wonder: Is this an established, fixed schema, or can it still be changed? -- If the latter, I'd like to propose some tweaks.

**Describe the solution you'd like**

Here is a short, pretty-printed example of the current format:

```json
{
  "xmin": 0,
  "xmax": 0.64,
  "tiers": [
    {
      "class": "IntervalTier",
      "name": "words",
      "xmin": 0.0,
      "xmax": 0.64,
      "entries": [
        [0.0, 0.07, "è"],
        [0.07, 0.62, "chiusa"],
        [0.62, 0.64, ""]
      ]
    },
    {
      "class": "IntervalTier",
      "name": "phones",
      "xmin": 0.0,
      "xmax": 0.64,
      "entries": [
        [0.0, 0.07, "ɛ"],
        [0.07, 0.2, "k"],
        [0.2, 0.25, "j"],
        [0.25, 0.36, "u"],
        [0.36, 0.45, "z"],
        [0.45, 0.62, "a"],
        [0.62, 0.64, ""]
      ]
    }
  ]
}
```

Here's how I propose to represent the same data:

```json
{
  "tmin": 0,
  "tmax": 0.64,
  "tiers": {
    "words": {
      "tmin": 0.0,
      "tmax": 0.64,
      "entries": [
        [0.0, 0.07, "è"],
        [0.07, 0.62, "chiusa"],
        [0.62, 0.64, ""]
      ]
    },
    "phones": {
      "tmin": 0.0,
      "tmax": 0.64,
      "entries": [
        [0.0, 0.07, "ɛ"],
        [0.07, 0.2, "k"],
        [0.2, 0.25, "j"],
        [0.25, 0.36, "u"],
        [0.36, 0.45, "z"],
        [0.45, 0.62, "a"],
        [0.62, 0.64, ""]
      ]
    }
  }
}
```

Here's my reasoning:

* In your schema, `tiers` is an array. This means that code desiring to extract just the words or just the phones must iterate over it, searching for the entry with the desired name. By modeling `tiers` as an object instead, selecting a given tier is as easy as accessing `tiers.words`, while still maintaining the possibility of iterating over all tiers.
* The tier objects are nested under a `tiers` property, so their role is clear from their position. The property `"class": "IntervalTier"` just adds noise, so I'd omit it.
* To me, the "x" in `xmin` and `xmax` sound like we're talking about spatial data. Since it's temporal data, I'd find `tmin` and `tmax` more intuitive.

That said, let me stress how happy I am that I can now instruct MFA to create JSON instead of TextGrid data. Regardless of schema details, this feature makes my life much easier!

**Comment by mmcauliffe** [comment]
Taking a look at this now since I've been playing with some other export formats.  The current schema is based on praatio's json format, which is just a [json dump of how it represents TextGrids internally](https://github.com/timmahrt/praatIO/blob/50dbff6efb33a93c675d993b6deb08505e45e094/praatio/utilities/textgrid_io.py#L347).  JSONs themselves are not supported by Praat, so there's no compatibility concerns there, but I would like for the JSONs exported by MFA to be readable by other programs.  

That said, I think probably the clearest format and least redundant format would be something like:

```
{
  "min_time": 0,
  "max_time": 0.64,
  "tiers": {
    "words": {
      "type": "interval",
      "entries": [
        [0.0, 0.07, "è"],
        [0.07, 0.62, "chiusa"],
        [0.62, 0.64, ""]
      ]
    },
    "phones": {
      "type": "interval",
      "entries": [
        [0.0, 0.07, "ɛ"],
        [0.07, 0.2, "k"],
        [0.2, 0.25, "j"],
        [0.25, 0.36, "u"],
        [0.36, 0.45, "z"],
        [0.45, 0.62, "a"],
        [0.62, 0.64, ""]
      ]
    }
  }
}
```

The changes on top of you @DanielSWolf 's changes would be:

1. Renaming "tmin" and "tmax" to "min_time" and "max_time", since there's no reason to be beholden to the "xmin" format that Praat uses
2. Removing "tmin" and "tmax" from the tiers (since I haven't come across a case where they differ for good reasons, it's more because in Praat an interval doesn't have access to any facts about the TextGrid it belongs to, but I could be wrong about use cases with different min/max times).
3. Adding "type" fields to tiers (MFA only uses interval tiers, but there are point tiers in Praat more generally)

@timmahrt have you thought about a more concise and transparent JSON schema for TextGrids?

**Comment by timmahrt** [comment]
Thank you for the ping!

> Removing "tmin" and "tmax" from the tiers (since I haven't come across a case where they differ for good reasons, it's more because in Praat an interval doesn't have access to any facts about the TextGrid it belongs to, but I could be wrong about use cases with different min/max times).

I haven't come across a use case either although it is technically possible. I can imagine that somewhere, someone is using tier-specific tmin and tmax but I'm ok with raising an exception when the tier's tmin/tmax do not equal the textgrid's tmin/tmax specifically when exporting to json.

----

The only unmentioned concern is one item with regards to compatibility with Praat.  In Praat, the list of tiers is ordered--thus the array.

So I'd add a parameter to the root level
"tier_name_list" which stores the ordered names of tiers.

With that we should be able to map between json and TextGrid file types losslessly.

[As an aside: Praat does not require tier names to be unique--I guess tiers are accessed positionally.  That would be problematic for moving away from the list of tiers to a tier hash except that praatio **does** require tier names to be unique.  🙈 ]

**Comment by DanielSWolf** [comment]
That all sounds good to me! Some thoughts:

* Due to its roots in JavaScript, JSON identifiers are typically in camel case. So things like `min_time`, `max_time`, and `tier_name_list` might more idiomatically be called `minTime`, `maxTime`, and `tierNameList`, respectively.
* For conciseness, we could just use `start` and `end` instead of `minTime` and `maxTime`.
* I didn't realize the tiers had to be sorted. If that's a requirement, then I'd opt to model them as an array after all. That way, my "find tier by name" scenario gets slightly harder, but the "treat tiers as sorted list" scenario gets much easier.

Here's what that might look like, based on @mmcauliffe's last changes:

```json
{
  "start": 0,
  "end": 0.64,
  "tiers": [
    {
      "name": "words",
      "type": "interval",
      "entries": [
        [0.0, 0.07, "è"],
        [0.07, 0.62, "chiusa"],
        [0.62, 0.64, ""]
      ]
    },
    {
      "name": "phones",
      "type": "interval",
      "entries": [
        [0.0, 0.07, "ɛ"],
        [0.07, 0.2, "k"],
        [0.2, 0.25, "j"],
        [0.25, 0.36, "u"],
        [0.36, 0.45, "z"],
        [0.45, 0.62, "a"],
        [0.62, 0.64, ""]
      ]
    }
  ]
}
```

**Comment by timmahrt** [comment]
> I didn't realize the tiers had to be sorted. If that's a requirement, then I'd opt to model them as an array after all. That way, my "find tier by name" scenario gets slightly harder, but the "treat tiers as sorted list" scenario gets much easier.

For the purpose of accessing the data, people usually don't need the order information--accessing by name is much more natural I think.  But I'm ok either way.

**Comment by mmcauliffe** [comment]
Ok so for the next release I'll have json output with the following schema:

```{
  "start": 0,
  "end": 0.64,
  "tiers": {
    "words": {
      "type": "interval",
      "entries": [
        [0.0, 0.07, "è"],
        [0.07, 0.62, "chiusa"],
        [0.62, 0.64, ""]
      ]
    },
    "phones": {
      "type": "interval",
      "entries": [
        [0.0, 0.07, "ɛ"],
        [0.07, 0.2, "k"],
        [0.2, 0.25, "j"],
        [0.25, 0.36, "u"],
        [0.36, 0.45, "z"],
        [0.45, 0.62, "a"],
        [0.62, 0.64, ""]
      ]
    }
  }
}
```

* Renamed `min_time` and `max_time` to `start` and `end`
* Kept tiers as a dictionary, because I think it's generally more useful to access them, and I *believe* that loading a dictionary from json in python will keep the order since the default dictionary is ordered, at least for iterating over it.  Accessing it will have to be by name, but that feels more intuitive to me too

**Comment by timmahrt** [comment]
If you prefer, I can make the changes inside of praatio.  They seem fairly useful for a general audience.

**Comment by mmcauliffe** [comment]
Yeah that'd be great!  I can cut back over to just passing "json" to `TextGrid.save` at some point.

**Comment by douglas125** [comment]
For the record, this is a good change (though I just realized it breaks my code that used the previous version LOL)
